### PR TITLE
Sockets added

### LINKED ISSUE #4: Sockets: making messaging real-time
- Add socket support on the front-end and back-end so that the messages are real-time 
- use this package socket.io for real-time notifications
- Setup socket server on the backend
- Setup client connection (should connect if the user is authed, add some layer of security would be best)
- If you are planning on doing online statuses, we also want an object to keep track of which users are currently connected - for simplicity, you can for now just use an in-memory variable (like a dictionary or object in the back-end - not recommended for scalability, but it works for this project)

### PR DESCRIPTION
**Pull request by metlis**
Closes #4 

The demonstration of how live messaging and statuses work;
Peek 2021-04-18 00-41

### CODE REVIEW

File: client/src/components/dashboard/Sidebar.jsx [hunk 1]
```diff
@@ -34,7 +34,7 @@ export default function Sidebar(props) {
       )
     });
   } else {
-    panels = conversations.map(conv => {
+    panels = conversations.filter(i => i.total_messages > 0).map(conv => {
```
Comment (WhoaNellie) [inline client/src/components/dashboard/Sidebar.jsx:37]: Could this filter potentially be done above on line 24 to reduce the number of iterations through the conversations?

**Review by WhoaNellie — COMMENTED** [review]
Great job on controlling when the socket server is initialized and connected/disconnected!

Just had one question about possibly optimizing getting users' conversations since that's an area where efficiency could be tricky

File: client/src/store/conversations.js [hunk 2]
```diff
@@ -31,11 +32,31 @@ export const conversationsSlice = createSlice({
       }
     },
     addMessage: (state, action) => {
-      state.active.messages.push(action.payload);
-      const conversation = state.list.find(i => i.id === state.active.id);
+      const toActive = state.active.id === action.payload.conversation;
+      if (toActive) state.active.messages.push(action.payload.message);
+
+      const conversation = state.list.find(i => i.id === action.payload.conversation);
```
Comment (ThinkTink) [inline client/src/store/conversations.js:38]: One optimization would be to check if `toActive`, and if so just set conversation to `state.active`, right?

File: client/src/store/conversations.js [hunk 3]
```diff
@@ -31,11 +32,31 @@ export const conversationsSlice = createSlice({
...
+    updateUserStatus: (state, action) => {
+      state.list.forEach(i => {
+        const user = i.users.find(u => u.id === action.payload.id);
+        if (user) {
+          user.active = action.payload.active;
+        }
+      })
+      const user = (state.active.users || []).find(i => i.id === action.payload.id);
+      if (user) {
+        user.active = action.payload.active;
```
Comment (ThinkTink) [inline client/src/store/conversations.js:59]: We're doing an awful lot of looping overall. I don't think anything needs to be changed at this point, but something to think about in the future is how to design how our data is stored to maximize efficiency

File: server/api/conversation_handler.py [hunk 4]
```diff
@@ -57,6 +67,14 @@ def messages(conversation_id):
         message = Message(conversation_id=conversation.id, author_id=current_user.id, text=text)
         db.session.add(message)
         db.session.commit()
+        socketio.emit(
```
Comment (ThinkTink) [inline server/api/conversation_handler.py:70]: we should also check against active_users here, right?

File: server/app.py [hunk 5]
```diff
@@ -22,12 +23,40 @@
...
 
 @login_manager.user_loader
 def get_user(user_id):
     return User.query.get(user_id)
+
+
+@socketio.on('connect')
+def connect():
+    if current_user.is_authenticated:
+        active_users[current_user.id] = request.sid
```
Comment (ThinkTink) [inline server/app.py:50]: Doing it this way limits a user to one connected device being active at a time, but someone might want to be able to use multiple, say, desktop and phone, simultaneously

**Review by ThinkTink — APPROVED** [review]
nice work overall, this is a very clean implementation of sockets. I've suggested a couple of tweaks, and then this is good to merge!

File: client/src/store/conversations.js (hunk 2 above)
Comment (metlis) [reply client/src/store/conversations.js:42]: optimized this part

File: client/src/store/conversations.js (hunk 3 above)
Comment (metlis) [reply client/src/store/conversations.js]: reduced looping a bit, but I agree that it still needs refactoring

File: server/api/conversation_handler.py (hunk 4 above)
Comment (metlis) [reply server/api/conversation_handler.py:75]: in this case the room for conversation should already exist, so there is no need in this check

**Pull request merged**